src/parser/CodeBlock.h

```diff
@@ -63,34 +63,159 @@ struct NativeFunctionInfo {
+        : m_fn(fn)
+    {
+    }
+
     NativeFunctionPointer m_fn;
 };
 
+class NativeCodeBlock;
 class InterpretedCodeBlock;
 
 class CodeBlock : public gc {
+public:
+    static const unsigned MAX_FUNCTION_LENGTH = 65536;
+
+    void* operator new(size_t size) = delete;
+    void* operator new[](size_t size) = delete;
+
+    virtual uint16_t functionLength() const = 0;
+    virtual AtomicString functionName() const = 0;
+
+    virtual bool isNativeCodeBlock() const
+    {
+        return false;
+    }
+
+    virtual bool isInterpretedCodeBlock() const
+    {
+        return false;
+    }
+
+    NativeCodeBlock* asNativeCodeBlock()
+    {
+        ASSERT(isNativeCodeBlock());
+        return (NativeCodeBlock*)this;
+    }
+
+    InterpretedCodeBlock* asInterpretedCodeBlock()
+    {
+        ASSERT(isInterpretedCodeBlock());
+        return (InterpretedCodeBlock*)this;
+    }
+
+    Context* context()
+    {
+        return m_context;
+    }
+
+    bool isStrict() const
+    {
+        return m_isStrict;
+    }
+
+protected:
+    CodeBlock()
+        : m_context(nullptr)
+    {
+    }
+
+    CodeBlock(Context* ctx, bool isStrict)
+        : m_context(ctx)
+        , m_isStrict(isStrict)
+    {
+    }
+
+    Context* m_context;
+    bool m_isStrict;
+};
+
+class NativeCodeBlock : public CodeBlock {
+    friend int getValidValueInNativeCodeBlock(void* ptr, GC_mark_custom_result* arr);
+
+public:
+    void* operator new(size_t size);
+    void* operator new[](size_t size) = delete;
+
+    NativeCodeBlock(Context* ctx, const NativeFunctionInfo& info)
+        : CodeBlock(ctx, info.m_isStrict)
+        , m_isNativeConstructor(info.m_isConstructor)
+        , m_length(info.m_argumentCount)
+        , m_name(info.m_name)
+        , m_nativeFunctionData((CallNativeFunctionData*)(new (PointerFreeGC) CallNativeFunctionData(info.m_nativeFunction)))
+    {
+        ASSERT(info.m_argumentCount < MAX_FUNCTION_LENGTH);
+    }
+
+    NativeCodeBlock(Context* ctx, const NativeFunctionInfo& info, CallNativeFunctionData* nativeData)
+        : CodeBlock(ctx, info.m_isStrict)
+        , m_isNativeConstructor(info.m_isConstructor)
+        , m_length(info.m_argumentCount)
+        , m_name(info.m_name)
+        , m_nativeFunctionData(nativeData)
+    {
+        ASSERT(info.m_argumentCount < MAX_FUNCTION_LENGTH);
+    }
+
+    virtual bool isNativeCodeBlock() const override
+    {
+        return true;
+    }
+
+    virtual uint16_t functionLength() const override
+    {
+        return m_length;
+    }
+
+    virtual AtomicString functionName() const override
+    {
+        return m_name;
+    }
+
+    bool isNativeConstructor() const
+    {
+        return m_isNativeConstructor;
+    }
+
+    CallNativeFunctionData* nativeFunctionData()
+    {
+        return m_nativeFunctionData;
+    }
+
+private:
+    bool m_isNativeConstructor;
+    uint16_t m_length;
+    AtomicString m_name;
+    CallNativeFunctionData* m_nativeFunctionData;
```

Comment:
IMO We can store CallNativeFunctionData directly.
it can reduce gc_malloc count & memory usage

Reply:
Yes, you catched the good point!
But `CallNativeFunctionData` is used also for api function and template code with additional function pointers.
IMO child class of `NativeFunctionObject` needs to be added to handle these extended cases.
Or if you have any other idea I'm open to it.
(Since this patch already contains a lot of changes, I'll refactor it in the next patch)